### PR TITLE
EVG-2545 Expansions for Patches created by Github Pull Requests

### PR DESCRIPTION
**Pull request by richardsamuels**
I've already tested this in staging.

### CODE REVIEW

File: agent/task.go [hunk 1]
```diff
@@ -228,8 +229,23 @@ func (a *Agent) getTaskConfig(ctx context.Context, tc *taskContext) (*model.Task
 		return nil, errors.New("agent retrieved an empty project ref")
 	}
 
+	var confPatch *patch.Patch
+	if confVersion.Requester == evergreen.GithubPRRequester {
+		tc.logger.Execution().Info("Fetching patch.")
+		confPatch, err = a.comm.GetTaskPatch(ctx, tc.task)
+		if err != nil {
+			return nil, err
```
Comment (bsamek) [inline agent/task.go]: wrap err?

File: agent/task.go [hunk 2]
```diff
@@ -228,8 +229,23 @@ func (a *Agent) getTaskConfig(ctx context.Context, tc *taskContext) (*model.Task
...
+		}
+		if confPatch == nil {
+			return nil, errors.New("agent retrieved an empty patch")
+		}
+
+	} else {
```
Comment (bsamek) [inline agent/task.go]: don't log in the else case

File: agent/task.go [hunk 3]
```diff
@@ -228,8 +229,23 @@ func (a *Agent) getTaskConfig(ctx context.Context, tc *taskContext) (*model.Task
...
 
+	var confPatch *patch.Patch
+	if confVersion.Requester == evergreen.GithubPRRequester {
+		tc.logger.Execution().Info("Fetching patch.")
+		confPatch, err = a.comm.GetTaskPatch(ctx, tc.task)
+		if err != nil {
+			return nil, err
+		}
+		if confPatch == nil {
+			return nil, errors.New("agent retrieved an empty patch")
```
Comment (bsamek) [inline agent/task.go]: what would cause confPatch to be nil and err to not be nil? will we ever hit this case?

File: agent/task.go (hunk 1 above)
Comment (bsamek) [inline agent/task.go]: log to Execution in the err case

File: agent/task.go [hunk 4]
```diff
@@ -228,8 +229,23 @@ func (a *Agent) getTaskConfig(ctx context.Context, tc *taskContext) (*model.Task
 		return nil, errors.New("agent retrieved an empty project ref")
 	}
 
+	var confPatch *patch.Patch
+	if confVersion.Requester == evergreen.GithubPRRequester {
+		tc.logger.Execution().Info("Fetching patch.")
```
Comment (bsamek) [inline agent/task.go]: "Fetching patch for GitHub pull request"

File: agent/task.go (hunk 3 above)
Comment (richardsamuels) [reply agent/task.go]: I changed the client method to generate an error when the API doesn't respond 200 OK, so it's now impossible. I've also removed this code.

File: agent/task.go (hunk 2 above)
Comment (richardsamuels) [reply agent/task.go]: done

File: agent/task.go (hunk 1 above)
Comment (richardsamuels) [reply agent/task.go]: done

File: agent/task.go (hunk 4 above)
Comment (richardsamuels) [reply agent/task.go]: done

File: rest/client/methods.go [hunk 5]
```diff
@@ -332,11 +332,14 @@ func (c *communicatorImpl) GetTaskPatch(ctx context.Context, taskData TaskData)
...
 	info.setTaskPathSuffix("git/patch")
 	resp, err := c.retryRequest(ctx, info, nil)
-
 	if err != nil {
 		return nil, errors.Wrapf(err, "could not get patch for %s", taskData.ID)
 	}
+	defer resp.Body.Close()
 
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.Errorf("could not find patch for task with id '%s'", taskData.ID)
```
Comment (bsamek) [inline rest/client/methods.go]: I'd be more explicit here that the reason we're erroring is because we didn't get a 200 OK.

Comment (richardsamuels) [reply rest/client/methods.go]: yep, added the status code.

Comment (bsamek) [reply rest/client/methods.go]: lgtm

File: model/task_config.go [hunk 6]
```diff
@@ -22,7 +23,7 @@ type TaskConfig struct {
 	WorkDir      string
 }
 
-func NewTaskConfig(d *distro.Distro, v *version.Version, p *Project, t *task.Task, r *ProjectRef) (*TaskConfig, error) {
+func NewTaskConfig(d *distro.Distro, v *version.Version, p *Project, t *task.Task, r *ProjectRef, patchDoc *patch.Patch) (*TaskConfig, error) {
```
Comment (tychoish) [inline model/task_config.go]: can't you get the patch document from the task document? if so, why do we need to pass it in?

Comment (richardsamuels) [reply model/task_config.go]: I'm not aware of any way to get the patch document from the task. Is there a specific method you're thinking of?

Comment (tychoish) [reply model/task_config.go]: somehow, the API is doing it in GetTaskPatch, so I presume its possible, presumably (though I'm not sure,) the version on the task is the patchid

**Comment by richardsamuels** [comment]
@tychoish Added a test for populateExpansions, GetTaskPatch sends all the task data, and the API is fetching patch by looking up patch by the Version field.

**Pull request merged**